clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -181,6 +215,15 @@ public synchronized void update(Cluster cluster, long now) {
     }
 
     /**
+     * If expiry is enabled, remove the unknown topic from the metadata refresh set to
+     * avoid further metadata requests for deleted topics which are not in use.
+     */
+    public synchronized void handleUnknownTopic(String topic) {
```

Comment:
Yes, the `cluster.nodes.size() > 0` check is annoying and ad-hoc. It could happen if there are no alive brokers (but then what broker is responding?) or if `MetadataResponse.getErrorResponse` is called.

My main concern with the current approach is that we trade an atomic `update` with something that is harder to reason about from a concurrency perspective.

Would the right solution be to have `update` and a `handleFailedUpdate` or something like that?